listener.go

```diff
@@ -0,0 +1,126 @@
+package main
+
+import (
+	"log"
+	"net"
+
+	baps3 "github.com/UniversityRadioYork/baps3-go"
+)
+
+// Maintains communications with the connector and connected clients.
+// Also does any processing needed with the commands.
+type hub struct {
+	// All current connections, and their outbound channel.
+	clients map[net.Conn]chan<- baps3.Message
+
+	// For communication with the connector.
+	cReqCh chan<- baps3.Message
+	cResCh <-chan baps3.Message
+
+	// Where new requests from clients come through.
+	reqCh chan baps3.Message
+
+	// Handlers for adding/removing connections.
+	addCh chan *Client
+	rmCh  chan *Client
+}
+
+var h = hub{
+	clients: make(map[net.Conn]chan<- baps3.Message),
+
+	reqCh: make(chan baps3.Message),
+
+	addCh: make(chan *Client),
+	rmCh:  make(chan *Client),
+}
+
+// Handles a new client connection.
+// conn is the new connection object.
+func (h *hub) handleNewConnection(conn net.Conn) {
+	defer conn.Close()
+	client := &Client{
+		conn:  conn,
+		resCh: make(chan baps3.Message),
+		tok:   baps3.NewTokeniser(),
+	}
+
+	// Register user
+	h.addCh <- client
+
+	go client.Read(h.reqCh, h.rmCh)
+	client.Write(client.resCh, h.rmCh)
+}
+
+func makeWelcomeMsg() *baps3.Message {
+	return baps3.NewMessage(baps3.RsOhai).AddArg("listd").AddArg("0.0")
```

Comment:
According to [the spec], this should only have _one_ argument, so something like 'listd-0.0'.  (It doesn't seem to say anything about how the version should be formatted, but this seems sensible.)

[the spec]: http://universityradioyork.github.io/baps3-spec/comms/internal/core.html#__code_ohai_code_server_welcome